worker/machineactions/worker.go

```diff
@@ -111,68 +111,92 @@ func (h *handler) Handle(abort <-chan struct{}, actionsSlice []string) error {
 		action, err := h.config.Facade.Action(actionTag)
 		if err != nil {
-			return errors.Annotatef(err, "could not retrieve action %s", actionId)
+			// If there is an error attempting to get the action, then don't bounce
+			// the worker. We can't remove the action notification directly, as that
+			// requires the action to exist.
+			// TODO (stickupkid) As a follow up, we should have a new method that
+			// allows the removal of a action notification without an action present.
+			logger.Infof("unable to retrieve action %s: %v", actionId, err)
+			continue
 		}
 
 		// Acquire concurrency slot.
-		h.limiter <- struct{}{}
+		select {
+		case h.limiter <- struct{}{}:
+		case <-abort:
+			// The associated strings watcher has been aborted, so there isn't
+			// anything we can do here but give up.
+			logger.Debugf("action %q aborted waiting in queue", actionTag.ID)
+			return nil
+		}
 		h.wait.Add(1)
-		go func(action machineactions.Action) {
-			var results map[string]interface{}
-			var actionErr error
-			defer func() {
-				// The result returned from handling the action is sent through using ActionFinish.
-				var finishErr error
-				if actionErr != nil {
-					finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionFailed, nil, actionErr.Error())
-				} else {
-					finishErr = h.config.Facade.ActionFinish(actionTag, params.ActionCompleted, results, "")
-				}
-				if finishErr != nil &&
-					!params.IsCodeAlreadyExists(finishErr) &&
-					!params.IsCodeNotFoundOrCodeUnauthorized(finishErr) {
-					logger.Errorf("could not finish action %s: %v", action.Name(), finishErr)
-				}
-
-				// Release concurrency slot.
-				<-h.limiter
-				h.wait.Done()
-			}()
-
-			if !action.Parallel() || action.ExecutionGroup() != "" {
-				group := "exec-command"
-				worker := "machine exec command runner"
-				if g := action.ExecutionGroup(); g != "" {
-					group = fmt.Sprintf("%s-%s", group, g)
-					worker = fmt.Sprintf("%s (exec group=%s)", worker, g)
-				}
-				spec := machinelock.Spec{
-					Cancel:  abort,
-					Worker:  worker,
-					Comment: fmt.Sprintf("action %s", action.ID()),
-					Group:   group,
-				}
-				releaser, err := h.config.MachineLock.Acquire(spec)
-				if err != nil {
-					actionErr = errors.Annotatef(err, "could not acquire machine execution lock for exec action %s", action.Name())
-					return
-				}
-				defer releaser()
-			}
-			err = h.config.Facade.ActionBegin(actionTag)
-			if err != nil {
-				actionErr = errors.Annotatef(err, "could not begin action %s", action.Name())
-				return
-			}
-			results, actionErr = h.config.HandleAction(action.Name(), action.Params())
-		}(*action)
+
+		// Run the action.
+		go h.runAction(actionTag, *action, abort)
 	}
 	return nil
 }
 
 // TearDown is part of the watcher.NotifyHandler interface.
 func (h *handler) TearDown() error {
 	// Wait for any running actions to finish.
+	// TODO (stickupkid): This wait group could wait for ever if any of actions hang.
```

Comment:
This was related to the machine wanting to restart. If the queue is full (100 actions running a the same time) and are blocked, then this will prevent the worker from restarting.

Reply:
Essentially there is a very basic semaphore that doesn't behave on shut down

Reply:
I think for now, just a simple log "Waiting for running actions..." then a "done". should be enough to see if we are blocking on this in production.